split up eslint config, added lerna

stefan-wild opened the pull request:
resolves #1 

the eslint config is now split up into seperate packages, and they extend one another 
`@madebywild/eslint-config-base` >  `@madebywild/eslint-config-typescript ` > `@madebywild/eslint-config`

and added lerna/github config based on https://github.com/madebywild/github-actions/tree/main/packages/lerna-release

for it to fully work I think it still needs:
- [x] find a way to test without public release?
- [x] update readme (maybe after #2 ?)
- [x] add secret for github publish action to npm
- [x] lint monorepo with latest`@madebywild/eslint-config`

stefan-wild committed:
split up eslint config, added lerna, bumped version to 0.2.3

r0skar commented on packages/eslint-config-base/index.js:
```diff
@@ -0,0 +1,30 @@
+module.exports = {
+  extends: ["plugin:prettier/recommended"],
+  plugins: ["prettier"],
+  overrides: [],
```
Imo we should remove the empty array.

r0skar commented on packages/eslint-config-typescript/index.js:
```diff
@@ -0,0 +1,25 @@
+module.exports = {
+  extends: ["@madebywild/eslint-config-base"],
+  overrides: [
+    {
+      files: ["**/*.js?(x)"],
+      rules: {},
+    },
```
Imo we should remove this. Any JS rule should be part of the base config.

r0skar commented on package.json:
```diff
@@ -1,44 +1,22 @@
-    "type": "git",
-    "url": "git+https://github.com/madebywild/eslint-config.git"
-  },
+  "name": "@madebywild/eslint-configs",
```
I guess the plural is needed because we now have a react-config that is named `eslint-config`? Maybe we should consider renaming that one to `eslint-config-react`? cc @lorenzomigliorero

lorenzomigliorero replied:
oh well, the main config isn’t used, it is just a monorepo requirement. The name can be simply changed as `eslint-config`, no matter if it's equal to others packages.
I would prefer to use just `@madebywild/eslint-config` for the main config, as other popular libs.
Otherwise, we’ll have another `@madebywild/eslint-*` distributed on NPM.

r0skar commented on packages/eslint-config-typescript/index.js:
```diff
@@ -0,0 +1,25 @@
+      files: ["**/*.js?(x)"],
+      rules: {},
+    },
+    {
+      files: ["**/*.ts?(x)"],
+      rules: {
+        "@typescript-eslint/no-explicit-any": "warn",
+        "@typescript-eslint/no-empty-function": "warn",
+        "@typescript-eslint/no-unused-vars": [
+          "warn",
+          {
+            ignoreRestSiblings: true,
+            args: "none",
+            varsIgnorePattern: "^_",
+          },
+        ],
+      },
+    },
```
We can move this to the top-level `rules`object now I think.

stefan-wild replied:
@r0skar I wasn't sure if that would also interfere with js files in a typescript project, but I guess that shouldn't happen anyway (or could be me just remembering less nice linting from a couple years ago)

r0skar replied:
Oh, good point. I didnt think about that. You are right that the rules would apply to JS files. On one hand I think some of the rules (eg. no-empty-function) can be useful for JS files as well. But still, its probably a good idea to keep the override.

r0skar commented on packages/eslint-config/index.js:
```diff
@@ -0,0 +1,14 @@
+module.exports = {
+  extends: ["react-app", "@madebywild/eslint-config-typescript"],
+  overrides: [],
```
Imo we should remove the empty array.

r0skar reviewed (commented):
I noticed the `version` now differs for each package. Should we reset them all to 0.0.1 (at least for the new packages it makes sense to me)? Not sure about best practice here @lorenzomigliorero

lorenzomigliorero commented:
For test the configurations without releasing them on NPM, we can temporarily publish them to our private verdaccio https://docs.wild.plus/private-registry#what-about-our-npm-packages.

lorenzomigliorero commented:
@r0skar we can't reset the version for `@madebywild/eslint-config`, since it's already published under NPM.
At the end, we'll have:

`@madebywild/eslint-config` => 0.2.3
`@madebywild/eslint-config-base` => 0.1.0
`@madebywild/eslint-config-typescript` => 0.1.0

I've updated the DoD in the issue with the expected versions.

@stefan-wild we don't need to manually bump the version, it is something in charge of `lerna`.
When we finished with the rest, I can take a look at the "publish" staff.

lorenzomigliorero commented on packages/eslint-config-base/package.json:
```diff
@@ -0,0 +1,34 @@
+  "main": "index.js",
+  "dependencies": {
+    "babel-eslint": "^10.1.0",
+    "eslint": "^7.25.0",
+    "eslint-config-prettier": "^8.3.0",
+    "eslint-plugin-flowtype": "^5.7.2",
+    "eslint-plugin-import": "^2.22.1",
+    "eslint-plugin-prettier": "^3.4.0"
```
afaik, these should be listed as `peerDependencies`, like in other popular configs.
https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/package.json#L67-L69

lorenzomigliorero commented on packages/eslint-config/package.json:
```diff
@@ -0,0 +1,35 @@
+  "main": "index.js",
+  "dependencies": {
+    "@madebywild/eslint-config-typescript": "^0.2.3",
+    "eslint-config-react-app": "^6.0.0",
```
same of https://github.com/madebywild/eslint-config/pull/3/files#r682683295

lorenzomigliorero commented on packages/eslint-config/package.json:
```diff
@@ -0,0 +1,35 @@
+  "main": "index.js",
+  "dependencies": {
+    "@madebywild/eslint-config-typescript": "^0.2.3",
+    "eslint-config-react-app": "^6.0.0",
+    "eslint-plugin-jsx-a11y": "^6.4.1",
+    "eslint-plugin-react": "^7.23.2",
+    "eslint-plugin-react-hooks": "^4.2.0"
```
same of https://github.com/madebywild/eslint-config/pull/3/files#r682683295

stefan-wild commented:
@lorenzomigliorero I thought it would be cleaner if they all have a matching version
or do you think no one would look at it anyway?

stefan-wild committed:
removed empty overrides

stefan-wild committed:
eslint-config-typescript, ts only overrides are now rules

r0skar commented:
On a general note about the typescript rules. In the past we relied on the [create-react-app to set the parser for TS](https://github.com/facebook/create-react-app/blob/main/packages/eslint-config-react-app/index.js#L33). Now that the TS config is not extending CRA anymore, we probably want to add that part to the TS config (parser, ts-plugin) ourselves.

stefan-wild committed:
set eslint and plugins to peerDependencies

stefan-wild committed:
updated package versions of new packages

lorenzomigliorero commented:
@stefan-wild Yes, it could make sense to matches all the versions (it's enough to remove [this](https://github.com/madebywild/eslint-config/pull/3/files?file-filters%5B%5D=.js&file-filters%5B%5D=.json#diff-2d72bdead8afa0798d18995311992d684348a694c2d5e214e8e4d2b6153e4821R2) line).
My unique doubt is that we need to force the `1.0.0` as first publish, and I don't know if it is possible with lerna, I've to look at it.

stefan-wild committed:
eslint-config-typescript, moved typescript rules back into overrides

stefan-wild commented:
@lorenzomigliorero would it be possible to set up an action, that auto releases the packages to our private registry when there's a commit to a dev branch?

I remember (ts) eslint config to be kinda finicky to set up, and it would be nice to get it running quick

or release just release @madebywild/eslint-config-base there
(and I'll extend from that in a test project, just to check if the typescript parsing works there)

lorenzomigliorero commented:
Theoretically, when we'll merge on main, the configs will be published under our registry as you can see [here](https://github.com/madebywild/eslint-config/pull/3/files#diff-2d72bdead8afa0798d18995311992d684348a694c2d5e214e8e4d2b6153e4821R14).
(I added now the NPM_TOKEN secret)

If we want to test the configuration before publishing to NPM, we can simply remove the [branch restriction](https://github.com/madebywild/eslint-config/pull/3/files#diff-87db21a973eed4fef5f32b267aa60fcee5cbdf03c67fafdc2a9b553bb0b15f34R6) from the github-action, and on every commit (remember to follow the conventional commit prefix) a new version will be released on verdaccio.

When we're ready, we can remove the verdaccio conf, and manually publish to NPM (just the first time).

stefan-wild committed:
fix: prettier plugin loading and typescript parsing

lorenzomigliorero commented on packages/eslint-config-base/package.json:
```diff
@@ -23,6 +23,10 @@
   ],
   "main": "index.js",
+  "dependencies": {
+    "eslint-plugin-prettier": "^3.4.0",
+    "eslint-config-prettier": "^8.3.0"
```
Since they are used from all the configurations, we can flag them as `peerDependencies`.

lorenzomigliorero commented on packages/eslint-config-typescript/package.json:
```diff
@@ -25,9 +25,12 @@
   "dependencies": {
-    "@madebywild/eslint-config-base": "^0.2.3"
+    "@madebywild/eslint-config-base": "^0.1.0",
+    "eslint-config-prettier": "^8.3.0",
+    "eslint-plugin-prettier": "^3.4.0"
```
same of https://github.com/madebywild/eslint-config/pull/3/files/161a990a67f655b5b89e26095599d29aa3e882bb..49858376ee32a25e390dbfa58bb3fa5838a3d467#r685208094

stefan-wild replied:
I know, but that doesn't work, that's why it's in both spots

lorenzomigliorero commented on packages/eslint-config/package.json:
```diff
@@ -26,7 +26,9 @@
   "dependencies": {
-    "@madebywild/eslint-config-typescript": "^0.2.3"
+    "@madebywild/eslint-config-typescript": "^0.1.0",
+    "eslint-config-prettier": "^8.3.0",
```
same of https://github.com/madebywild/eslint-config/pull/3/files/161a990a67f655b5b89e26095599d29aa3e882bb..49858376ee32a25e390dbfa58bb3fa5838a3d467#r685208094

stefan-wild commented:
@lorenzomigliorero I tested the packages locally with npm link and by duplicating the index.js into .eslintrc.js
and the prettier things just don't work without being in dependencies (and the prettier dependencies also need to be installed in the project to work, at least for me in vscode)

and the stuff I found ends up like this
https://github.com/voxmedia/eslint-config-voxproduct/blob/master/package.json

maybe I'm missing something, so please feel free to see if you can find a nicer solution

r0skar commented on packages/eslint-config-typescript/index.js:
```diff
@@ -0,0 +1,21 @@
+module.exports = {
+  extends: ["@madebywild/eslint-config-base"],
+  parser: "@typescript-eslint/parser",
```
I think the parser belongs into the override as well. See https://github.com/facebook/create-react-app/blob/main/packages/eslint-config-react-app/index.js for an implementation example. (Also, we probably need the typescript plugin as well in the overrides?)

stefan-wild commented:
@lorenzomigliorero ok, my head is kinda smoking now from dealing with the dependecy/peerDependency stuff,
eslint folks want us to use peerDependecy, but that requires manually installing each dependecy in the project
https://github.com/eslint/eslint/issues/3458 (which seems pretty ugly to me)
or new npm, which seems to install peerDependecies by default https://github.blog/2020-10-13-presenting-v7-0-0-of-the-npm-cli/

lorenzomigliorero commented:
https://github.com/madebywild/eslint-config/blob/main/index.js is no more needed, we can delete it.

lorenzomigliorero commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
 node_modules
 package-lock.json
+yarn.lock
```
Why? We need the yarn.lock versioned in the repo.

lorenzomigliorero committed:
build: removed old index.js

lorenzomigliorero committed:
build: change root pkg name

lorenzomigliorero committed:
build: remove yarn.lock from gitignore

lorenzomigliorero committed:
Merge branch 'split-config' of github.com:madebywild/eslint-config into split-config

stefan-wild committed:
fix: set plugins

stefan-wild committed:
chore: moved everything to peerDependencies

stefan-wild committed:
Merge branch 'split-config' of github.com:madebywild/eslint-config into split-config

lorenzomigliorero committed:
ci: add registry info

stefan-wild committed:
fix: added stacked peerDependencies

lorenzomigliorero committed:
style: add eslint

lorenzomigliorero committed:
Merge branch 'split-config' of github.com:madebywild/eslint-config into split-config

lorenzomigliorero committed:
docs: add base documentation

stefan-wild committed:
docs: github README.md no points to docs.wild.plus docu and small updates

stefan-wild committed:
docs: updated elint-config readme.md

lorenzomigliorero committed:
ci: temporary removed branch restriction for release

lorenzomigliorero committed:
Merge branch 'split-config' of github.com:madebywild/eslint-config into split-config

lorenzomigliorero committed:
ci: add branch param

lorenzomigliorero committed:
ci: fix branch param syntax

lorenzomigliorero committed:
build: update yarn.lock

Commit:
chore(release): [skip ci]

 - @madebywild/eslint-config-base@0.1.1
 - @madebywild/eslint-config-typescript@0.1.1
 - @madebywild/eslint-config@0.2.4

stefan-wild committed:
docs: simplified peerdeps installing

Commit:
chore(release): [skip ci]

 - @madebywild/eslint-config-base@0.1.2
 - @madebywild/eslint-config-typescript@0.1.2
 - @madebywild/eslint-config@0.2.5

stefan-wild committed:
fix: removed wrong plugin

stefan-wild committed:
Merge branch 'split-config' of github.com:madebywild/eslint-config into split-config

Commit:
chore(release): [skip ci]

 - @madebywild/eslint-config-base@0.1.3
 - @madebywild/eslint-config-typescript@0.1.3
 - @madebywild/eslint-config@0.2.6

stefan-wild commented:
alright, just tried adding `@madebywild/eslint-config`(split one, from the private registry) to an existing project, and after running `npx install-peerdeps @madebywild/eslint-config-typescript --dev` it was working alright :D

@lorenzomigliorero do you want to check some more, or do you think this is completed now?

stefan-wild committed:
docs: added peer dependency install instructions for different npm versions

Commit:
chore(release): [skip ci]

 - @madebywild/eslint-config-base@0.1.4
 - @madebywild/eslint-config-typescript@0.1.4
 - @madebywild/eslint-config@0.2.7

stefan-wild committed:
ci: also ignore mdx changes

stefan-wild committed:
Merge branch 'split-config' of github.com:madebywild/eslint-config into split-config

Commit:
chore(release): [skip ci]

 - @madebywild/eslint-config-base@0.1.5
 - @madebywild/eslint-config-typescript@0.1.5
 - @madebywild/eslint-config@0.2.8

lorenzomigliorero commented:
@stefan-wild I think we can proceed with the NPM publication, I just need to figure out how obtain a token for the https://www.npmjs.com/org/madebywild organization.

lorenzomigliorero committed:
ci: removed private registry

lorenzomigliorero committed:
Merge branch 'split-config' of github.com:madebywild/eslint-config into split-config

lorenzomigliorero committed:
ci: re-added main branch restriction
